typing/btype.ml

```diff
@@ -1005,21 +1009,58 @@ module Alloc_mode = struct
        submode_exn (Amode v.upper) (Amodevar v);
        v.upper
 
+  exception Became_constant
   let compress_vlower v =
+    let nmarked = ref 0 in
+    let mark v' =
+      assert (not v'.mark);
+      v'.mark <- true;
+      incr nmarked
+    in
+    let unmark v' =
+      assert v'.mark;
+      v'.mark <- false;
+      decr nmarked
+    in
+    let log_head = ref Unchanged in
+    let log = ref log_head in
+    let new_vlower = ref v.vlower in
     (* Ensure that each transitive lower bound of v
        is a direct lower bound of v *)
     let rec trans v' =
       if le_const v'.upper v.lower then ()
-      else if List.memq v' v.vlower then ()
+      else if v'.mark then ()
       else begin
-        v.vlower <- v' :: v.vlower;
+        mark v';
+        new_vlower := v' :: !new_vlower;
         trans_low v'
       end
     and trans_low v' =
-      submode_exn (Amode v'.lower) (Amodevar v);
+      assert (v != v');
+      if not (le_const v'.lower v.upper) then
+        Misc.fatal_error "compress_vlower: invalid bounds";
+      if not (le_const v'.lower v.lower) then begin
+        let m = join_const v.lower v'.lower in
+        set_lower ~log v m;
```

Comment:
Since we're not editing `v.vlower` directly, it might be clearer to have a `new_lower` reference too, and make all the changes to `v` right at the end.

Reply:
Good idea, fixed.